ArduPlane/Attitude.cpp

```diff
@@ -5,31 +5,77 @@
   to PIDs to change the scaling of the PID with speed. At high speed
   we move the surfaces less, and at low speeds we move them more.
  */
-float Plane::calc_speed_scaler(void)
+Vector3f Plane::calc_rpy_speed_scaler(void)
 {
-    float aspeed, speed_scaler;
+    Vector3f speed_scaler;
+    float aspeed;
     if (ahrs.airspeed_estimate(aspeed)) {
         if (aspeed > auto_state.highest_airspeed && hal.util->get_soft_armed()) {
             auto_state.highest_airspeed = aspeed;
         }
+
+        Vector3f aspeed_rpy;
+        if (flight_stage != AP_FixedWing::FlightStage::VTOL && is_positive(g2.prop_disc_loading)) {
+            // Calculate prop-wash exit velocity using momentum disc theory
+            const float rho = SSL_AIR_DENSITY * plane.barometer.get_air_density_ratio();
+            // Estimate propeller disc loading in N/m^2 assuming drag is negligible and X axis
+            // specific force is mostly due to propeller thrust. This assumption will underestimate
+            // propeller disc loading at higher speeds but will be accurate enough at low speeds
+            // which is when prop wash effects dominate.
+            const float disc_loading = MAX(g2.prop_disc_loading * (ahrs.get_accel().x - ahrs.get_accel_bias().x), 0.0f);
+            // Solve this quadratic for propeller inflow factor a^2 + a - T / (2 * rho * S * V^2) = 0
+            // a = inflow factor
+            // T = propeller thrust (N)
+            // rho = air density (Kg/m^3)
+            // S = propeller area (m^2)
+            // V = true airspeed (m/s)
+            // Aerodynamics for Engineering Students, E L Houghton and N B Carruthers, Third Edition, pp469
+            const float a_coef = 1.0f;
+            const float b_coef = 1.0f;
+            const float c_coef = - disc_loading / (2.0f * rho * sq(MAX(aspeed,0.1f)));
+            const float inflow_factor = (-b_coef + sqrtf(sq(b_coef) - 4.0f * a_coef * c_coef)) / (2.0f * a_coef);
+            aspeed_rpy.x = aspeed * (1.0f + 2.0f * inflow_factor * g2.propwash_roll_comp);
```

Comment:
I think logging the inflow factor (or 2xinflow_factor?) would help with developing tools to find the right values for the compensation values from logs
I'm also a bit concerned about noise on the accel X axis, especially with IC planes. Should we filter it?

Reply:
I've added logging of the inflow factor. WRT the noise, the speed scaler does get filtered before it is used in Plane::calc_airspeed_errors() but I think it would be better to separately filter the airspeed and longitudinal acceleration before they are used in the prop wash velocity calculation or else large noise will be be rectified and result in an offset. I'll push a patch to split the filtering.